risc0/zkvm/src/prove/image.rs

```diff
@@ -0,0 +1,212 @@
+        let mem_size = PAGE_TABLE.start() + info.total;
+        let num_pages = mem_size / page_size;
+        for i in 0..num_pages {
+            let digest = sha.hash_bytes(&image[i * page_size..(i + 1) * page_size]);
+            let addr = PAGE_TABLE.start() + i * DIGEST_BYTES;
+            image[addr..addr + DIGEST_BYTES].copy_from_slice(digest.as_bytes());
+        }
+
+        // Now compute the final root hash.
+        let final_bytes = info.layers.last().unwrap();
+        let final_page_addr = PAGE_TABLE.start() + info.total - final_bytes;
+        let root = sha.hash_bytes(&image[final_page_addr..final_page_addr + final_bytes]);
+        let root_addr = PAGE_TABLE.start() + info.total;
+        image[root_addr..root_addr + DIGEST_BYTES].copy_from_slice(root.as_bytes());
+
+        Ok(Self {
+            image,
+            root: *root,
+            info,
+        })
+    }
+
+    pub fn check(&self, page_idx: usize) -> Result<()> {
+        let sha = sha::sha();
+
+        let mut page_idx = page_idx;
+        for _ in 0..self.info.layers.len() {
+            let addr = page_idx * self.info.page_size;
+            let page = &self.image[addr..addr + self.info.page_size];
+            let expected = sha.hash_bytes(page);
+            let page_table_addr = PAGE_TABLE.start() + page_idx * DIGEST_BYTES;
+            let page_table_entry = &self.image[page_table_addr..page_table_addr + DIGEST_BYTES];
+            let actual = Digest::from_bytes(page_table_entry);
+            if *expected != actual {
+                bail!("Invalid page table entry: {} != {}", *expected, actual);
+            }
+            page_idx = page_table_addr / self.info.page_size;
+        }
+
+        let final_bytes = self.info.layers.last().unwrap();
+        let final_page_addr = PAGE_TABLE.start() + self.info.total - final_bytes;
+        let final_page = &self.image[final_page_addr..final_page_addr + final_bytes];
+        let expected = sha.hash_bytes(final_page);
+        let root_addr = PAGE_TABLE.start() + self.info.total;
+        let root_bytes = &self.image[root_addr..root_addr + DIGEST_BYTES];
+        let actual = Digest::from_bytes(root_bytes);
+        if *expected != actual {
+            bail!("Invalid root hash: {} != {}", *expected, actual);
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use risc0_zkvm_methods::MULTI_TEST_ELF;
+    use risc0_zkvm_platform::{
+        memory::{PAGE_TABLE, PROG},
+        syscall::DIGEST_BYTES,
+    };
+
+    use super::MemoryImage;
+    use crate::prove::image::PageTableInfo;
+
+    fn page_table_size(max_mem: usize, page_size: usize) -> usize {
+        PageTableInfo::new(max_mem, page_size).total
+    }
+
+    #[test]
+    fn check_integrity() {
+        const PAGE_SIZE: usize = 1024;
+        let image = MemoryImage::new(MULTI_TEST_ELF, PAGE_SIZE).unwrap();
+        // std::fs::write("/tmp/test.img", &image.image).unwrap();
```

Comment:
This comment looks like it's left over from a previous draft. If that's the case, could you remove it?

Reply:
It's actually useful in case you want to examine the entire memory image.